src/background/tab-state.js

```diff
@@ -140,15 +143,47 @@ export default function TabState(initialState, onchange) {
   };
 
   /**
-   * Request the current annotation count for the tab's URL
+   * Request the current annotation count for the tab's URL.
+   * It debounces previous requests within the value defined by
+   * WAITMILLISECONDS
    *
    * @method
    * @param {integer} tabId The id of the tab.
    * @param {string} tabUrl The URL of the tab.
    */
   this.updateAnnotationCount = function (tabId, tabUrl) {
-    return uriInfo.getAnnotationCount(tabUrl).then(count => {
-      this.setState(tabId, { annotationCount: count });
+    const WAITMILLISECONDS = 3000; //
+    const pendingRequest = pendingAnnotationCountRequests.get(tabId);
+
+    // Cancel pending requests for the specific tabId, if any
+    if (pendingRequest && pendingRequest.cancel) {
+      pendingRequest.cancel();
+    }
+
+    const fetchCount = async () => {
+      try {
+        const annotationCount = await uriInfo.getAnnotationCount(tabUrl);
+        this.setState(tabId, { annotationCount });
+      } finally {
+        // Clear the pending request
+        pendingAnnotationCountRequests.delete(tabId);
+      }
+    };
+
+    return new Promise((resolve, reject) => {
+      const timerId = setTimeout(() => {
+        resolve(fetchCount());
+      }, WAITMILLISECONDS);
+
+      // Add the cancellation function that:
+      // 1. clears the timeout
+      // 2. rejects the promise
+      pendingAnnotationCountRequests.set(tabId, {
+        cancel: () => {
+          clearTimeout(timerId);
+          reject(new Error(`Canceled request of: ${tabUrl}`));
```

Comment:
I'm wary of rejecting `Promise`s in what is a normal branch of code (i.e. not an error condition). It may seem a little counter-intuitive, but I'd recommend resolving to `noValue`/`null` and letting the caller deal with that value accordingly.